File: src/types/scenario.ts
export type MitigationPose =
  | 'idle'
  | 'prepare'
  | 'duck'
  | 'cover'
  | 'hold'
  | 'crawl'
  | 'alert'
  | 'walk'
  | 'climb'
  | 'gather'
  | 'celebrate';

export type MitigationMood = 'normal' | 'scared';

export interface MitigationItem {
  id: string;
  text: string;
  note: string;
  pose: MitigationPose;
  mood: MitigationMood;
  quake?: boolean;
  tsunami?: boolean;
  flood?: boolean;
  fire?: boolean;
  moveX?: number;
  moveY?: number;
  color?: string;
}

export interface DisasterPhase {
  key: string; // sebelum | saat | sesudah
  label: string; // Pra-Bencana | Darurat | Pemulihan
  color: string;
  tint: string;
  icon: string;
  items: MitigationItem[];
}

export interface DisasterScenario {
  id: string;
  title: string;
  icon: string;
  scene: 'rumah' | 'sekolah' | 'pesisir' | 'banjir' | 'kebakaran';
  difficulty: string;
  description: string;
  phases: DisasterPhase[];
}

export interface ItemValidationResult {
  item_id: string;
  is_correct_position: boolean;
  is_correct_phase: boolean;
  expected_index: number;
  actual_index?: number | null;
  expected_phase: string;
}

export interface ExplanationItem {
  step_number: number;
  title: string;
  note: string;
  phase_label: string;
  phase_color: string;
}

export interface ValidationResponse {
  is_valid: boolean;
  all_filled: boolean;
  score_percentage: number;
  correct_count: number;
  total_count: number;
  results: ItemValidationResult[];
  explanations: ExplanationItem[];
  feedback_message: string;
}

export interface QuizQuestion {
  id: string;
  scenario_id: string;
  question: string;
  options: string[];
  correct_index: number;
  explanation: string;
  phase: string;
}

export interface QuizDetailedFeedback {
  question_id: string;
  question: string;
  user_answer: string;
  correct_answer: string;
  is_correct: string;
  explanation: string;
}

export interface QuizResult {
  user_name: string;
  score: int_or_number;
  total: int_or_number;
  percentage: number;
  passed: boolean;
  certificate_id?: string | null;
  date_issued: string;
  detailed_feedback: QuizDetailedFeedback[];
}

type int_or_number = number;

File: models/scenario.py
from datetime import datetime
from pydantic import BaseModel, Field
from typing import List, Dict, Optional


class MitigationItem(BaseModel):
    id: str
    text: str
    note: str
    pose: str
    mood: str
    quake: bool = False
    tsunami: bool = False
    flood: bool = False
    fire: bool = False
    moveX: int = 0
    moveY: int = 0
    color: Optional[str] = None


class DisasterPhase(BaseModel):
    key: str
    label: str
    color: str
    tint: str
    icon: str
    items: List[MitigationItem]


class DisasterScenario(BaseModel):
    id: str
    title: str
    icon: str
    scene: str
    difficulty: str
    description: str
    phases: List[DisasterPhase]


class ItemValidationResult(BaseModel):
    item_id: str
    is_correct_position: bool
    is_correct_phase: bool
    expected_index: int
    actual_index: Optional[int] = None
    expected_phase: str


class ExplanationItem(BaseModel):
    step_number: int
    title: str
    note: str
    phase_label: str
    phase_color: str


class ValidationRequest(BaseModel):
    scenario_id: str
    ordered_items: List[str]
    placed_items: Dict[str, str]


class ValidationResponse(BaseModel):
    is_valid: bool
    all_filled: bool
    score_percentage: float
    correct_count: int
    total_count: int
    results: List[ItemValidationResult]
    explanations: List[ExplanationItem]
    feedback_message: str


class ExplanationItem(BaseModel):
    step_number: int
    title: str
    note: str
    phase_label: str
    phase_color: str


class QuizQuestion(BaseModel):
    id: str
    scenario_id: str
    question: str
    options: List[str]
    correct_index: int
    explanation: str
    phase: str


class QuizSubmission(BaseModel):
    user_name: str
    answers: Dict[str, int]


class QuizResult(BaseModel):
    user_name: str
    score: int
    total: int
    percentage: float
    passed: bool
    certificate_id: Optional[str] = None
    date_issued: str
    created_at: datetime = Field(default_factory=datetime.utcnow)